src/solvers/mp/MasterProblem.h

```diff
@@ -0,0 +1,386 @@
+      return true;
+    }
+
+    // Returns true if both columns are disjoint PLUS ONE DAY INBETWEEN (needRest) !!
+    virtual bool isDisjointWith(PPattern pat, bool needRest=true) const {
+      return ((firstDay_+length_ < pat->firstDay_ - needRest)
+              || (pat->firstDay_+pat->length_ < firstDay_ - needRest)
+      );
+    };
+
+    // Returns true if both columns are disjoint PLUS ONE DAY INBETWEEN (needRest) !!
+    virtual bool isShiftDisjointWith(PPattern pat, bool needRest=true) const {
+      if(isDisjointWith(pat, needRest))
+        return true;
+
+      int commomFirstDay = std::max(firstDay_, pat->firstDay_),
+          commomLastDay=std::min(firstDay_+length_, pat->firstDay_+pat->length_)-1;
+      for(int k=commomFirstDay; k<=commomLastDay; ++k)
+        if(getShift(k) == pat->getShift(k)) return false;
+
+      return true;
+    };
+
+    virtual std::string toString(int nbDays = -1, std::vector<int> shiftIDToShiftTypeID={}) const {
+      return "";
+    }
+
+    virtual int getShift(int day) const = 0;
+
+    // need to be able to write the pattern as a vector and to create a new one from it
+    virtual std::vector<double> getCompactPattern() const {
+      std::vector<double> compact;
+      compact.push_back(nurseId_);
+      compact.push_back(firstDay_);
+      compact.push_back(length_);
+      return compact;
+    }
+
+    int nurseId_;
+    int firstDay_;
+    int length_;
+
+};
+
+//-----------------------------------------------------------------------------
+//
+//  C l a s s   M a s t e r P r o b l e m
+//
+// Build and solve the master problem of the column generation scheme
+//
+//-----------------------------------------------------------------------------
+
+enum MySolverType { S_SCIP, S_CLP, S_Gurobi, S_Cplex, S_CBC };
+static std::map<std::string,MySolverType> MySolverTypesByName =
+{{"CLP",S_CLP},{"Gurobi",S_Gurobi},{"Cplex",S_Cplex},{"CBC",S_CBC},{"SCIP",S_SCIP}};
+
+class MasterProblem : public Solver, public PrintSolution{
+  public:
+    // Specific constructor and destructor
+    MasterProblem(Scenario* pScenario, Demand* pDemand,
+      Preferences* pPreferences, std::vector<State>* pInitState, MySolverType solver);
+    ~MasterProblem();
+
+    //solve the rostering problem
+    double solve(std::vector<Roster> solution = {});
+
+    //solve the rostering problem or just the relaxation(root node)
+    double solve(std::vector<Roster> solution, bool rebuild);
+
+    // Solve with parameters
+    double solve(const SolverParam& parameters, std::vector<Roster> solution = {});
+
+    //Resolve the problem with another demand and keep the same preferences
+    //
+    double resolve(Demand* pDemand, const SolverParam& parameters, std::vector<Roster> solution = {});
+
+    // needs to be specialized: add a colum  to the master from a solution of the subproblem
+    virtual MyVar* addColumn(int nurseId, const RCSolution& solution) = 0;
+
+    // retrieve the object represented ny the  vector pattern
```

Comment:
The definition of a pattern implicitly assumes that the columns are rotations. Is it not best if moved subclass?

Reply:
I've used the same for roster too, so I think it's ok. Rotation implements Pattern.

Reply:
ok, makes sense